### ISSUE TITLE
DriveInfo constructor does not accept nonexistent path on Linux, only Windows

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/corefx#29524]
This code throws on Linux but works on Windows:
```
new DriveInfo(typeof(object).Assembly.Location + "\\doesnotexist").DriveFormat
```

This is because in DriveInfo.Windows.cs we do Path.GetPathRoot, but in DriveInfo.Linux.cs we do not.

We should be equally tolerant on Linux so that code ports smoothly.

Note the docs claim only the drive letter is accepted: I'll open an issue against docs.

**Comment by stephentoub** [comment]
> We should be equally tolerant on Linux so that code ports smoothly.

What do you expect it to do here?  On Windows it calls Path.GetPathRoot so that it can get the drive letter to match.  But on Linux, GetPathRoot will just return '/', which may not be the actual drive it's on.

**Comment by danmosemsft** [comment]
I can't think of a perfect solution. A "largest common substring" match against DriveInfos from GetDrives() is one possibility. Another is to repeatedly do Path.GetDirectoryName() until you find an existing one (if any). Both are string manipulation, they won't follow links of course.

Not an API I'd design, but may be better than code failing that we could potentially make somewhat useful on Unix.

**Comment by danmosemsft** [comment]
Even on Windows it doesn't follow junctions, it just cleaves the drive off the front. It's just less common to have junctions on Windows.

**Comment by MarcoRossignoli** [comment]
If i understood well here the solution is to change [NormalizeDriveName](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.DriveInfo/src/System/IO/DriveInfo.Unix.cs#L24) and try to looking for correct(if any) `DriveInfo`and return `DriveInfo.Name`, i'm not a linux expert but if this is not urgent i could try a PR.

**Comment by stephentoub** [comment]
> the solution is

I don't think we know what the right solution is, if we even want to do anything.

> A non related question, how do you choose if a code should be on coreclr or corefx?

We try to put things in corefx unless they need to be in coreclr, which is generally only necessary if something else in coreclr relies on it or is tightly integrated with it.

**Comment by MarcoRossignoli** [comment]
>if we even want to do anything

ok let me know...TYVM for clarifying.

**Comment by carlossanlop** [comment]
Triage:

This issue is a combination of at least these problems:

- [The documentation](https://docs.microsoft.com/en-us/dotnet/api/system.io.driveinfo.-ctor?view=netcore-3.1#System_IO_DriveInfo__ctor_System_String_) for the `DriveInfo` constructor states that the argument is `driveName`, which I think it means:
  - In Windows, we have to pass `C`, `C:` or `C:\` (they get resolved properly).
  - In Unix, we have to pass a drive from the `/dev/` folder, like `/dev/sda`, or `/dev/nvme0n1p5` (that's my case in my Ubuntu machine).
  - In Unix, the method works if you pass a real file or folder, but I don't think that is the real intention of the `DriveInfo` constructor - it's good intention from the P/Invoke that happens to accept file/folder paths and can return drive information.
- As Stephen commented, in Unix, the root of all files and folders is `/`, so `Path.GetPathRoot` is useless to retrieve drive information in the Unix case. `Path.GetPathRoot` would also be useless if the user does not pass a full path, because [it would return an empty string](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getpathroot?view=netcore-3.1).
- For the Unix case, the method requires the path to be a real, full path. We do not internally call `Path.GetFullPath` to [verify if the file can be prefixed](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfullpath?view=netcore-3.1#System_IO_Path_GetFullPath_System_String_) all the way to `/`. Even if we did call `Path.GetFullPath`, we would still have the problem described in the previous point: The full path does not give any drive information like in Windows.
- The `DriveInfo.DriveFormat` property (which is the one throwing the exception in this issue) internally makes a P/Invoke call to `SystemNative_GetFormatInfoForMountPoint`, which only works with _existing_ files. When the file does not exist, [the method returns -1 and then we throw](https://github.com/dotnet/runtime/blob/f40e72ff9484ebf9c681dae1e6f5d1dc88b58716/src/libraries/System.IO.FileSystem.DriveInfo/src/System/IO/DriveInfo.Unix.cs#L73).
- Since we can mount any drive in any subfolder in the filesystem, and, according to the description of this issue, we want to be able to pass a path for a file that does not yet exist, then at the very least _it's parent folder_ should exist. Example:
  - We could have `/home/user/mountedDriveA/FirstFile.txt` existing inside the `/dev/sda1` partition, but
  - We could  have `/home/user/mountedDriveB/SecondFile.txt` existing inside the `/dev/sdb1` partition.


I have some suggestions on how to fix this issue:

- We can modify `DriveFormat` so that before it makes the P/Invoke, it checks if the passed path is a file. If it is a file, we check if it exists. If it exists, we pass the path to the P/Invoke and it should succeed. If the file does not exist, we check if its folder exists. If it does not exist, we throw. If it exists, we pass the folder path to the P/Invoke, and it should succeed.

- Or we could call `Path.GetFullPath` inside the `DriveInfo` constructor. If the path does not exist, we throw.

- Or we could close this issue since the `DriveInfo` constructor should accept a drive/partition path, and we make sure to describe this more clearly in the documentation.

**Comment by carlossanlop** [comment]
Triage:
We had a conversation and a good option would be to add a new static method in the `DriveInfo` class, maybe named `GetDriveInfoForFullPath` that accepts a full path and finds to which drive that path belongs to, including the case of mounting a drive into a folder (which applies to both Windows and Unix).
